File: linked-list/MyLinkedList.java
// compile: javac <file>.java, file and class name must have same name
//      i.e. javac MyLinkedList.java  produce -> MyLinkedList.class
// run java <CompiledFile>.class
//      i.e. java MyLinkedList

public class MyLinkedList {
    public static void main(String args[]) {

        // String linked list
        LL<String> ll = new LL<>();

        ll.addLast("cat");
        ll.addLast("dog");
        ll.addFirst("alligator");

        ll.display();

        try {
            System.out.println("First Item: " + ll.getFirst());
            System.out.println("Last Item: " + ll.getLast());
        } catch (Exception e) {
            System.out.println(e);
        }


        // Integer linked list
        LL<Integer> ll2 = new LL<>();

        ll2.addLast(4);
        ll2.addLast(3);
        ll2.addLast(2);
        ll2.addLast(1);
        ll2.addFirst(5);
        ll2.addFirst(6);
        ll2.addFirst(7);

        ll2.removeFirst();
        ll2.removeLast();

        ll2.display();

        try {
            System.out.println("First Item: " + ll2.getFirst());
            System.out.println("Last Item: " + ll2.getLast());
        } catch (Exception e) {
            System.out.println(e);
        }
    }
}

class Node<V> {
    Node<V> next = null;
    V value;

    public Node(V val) {
        value = val;
    }
}

class LL<T> {
    private Node<T> head;

    public void LL() {
        head = null;
    }

    public void addFirst(T val) {
        Node<T> newNode = new Node<>(val);

        if (head == null) {
            head = newNode;
        } else {
            newNode.next = head;
            head = newNode;
        }
    }

    public void addLast(T val) {
        Node<T> newNode = new Node<>(val);

        if (head == null) {
            head = newNode;
        } else {
            Node<T> current = head;

            while (current != null && current.next != null) {
                current = current.next;
            }

            current.next = newNode;
        }
    }

    public void removeFirst() {
        if (head != null) {
            if (head.next != null) {
                head = head.next;
            } else {
                head = null;
            }
        }
    }

    public void removeLast() {
        if (head != null) {
            Node<T> current = head;

            while(current != null && current.next.next != null) {
                current = current.next;
            }

            current.next = null;
        }
    }

    public T getFirst() throws Exception {

        if (head == null) {
            throw new Exception("Empty linked list!");
        }

        return head.value;
    }

    public T getLast() throws Exception {
        Node<T> current = head;

        if (current == null) {
            throw new Exception("Empty linked list!");
        }

        while(current != null && current.next != null) {
            current = current.next;
        }

        return current.value;
    }

    public void display() {
        if (head != null) {
            Node<T> current = head;

            while(current != null) {
                System.out.print("[" + current.value + "]->");

                current = current.next;
            }

            System.out.println();
        }
    }
}

File: linked-list/MyLinkedList.ml
(* ocaml MyLinkedList.ml *)

(* Ocaml is functioal, imperative, and object oriented, can use the paradignm
   best fit for job *)

(* types can be inferred, but sometimes must be explicit *)

(* OOP impelementation *)
class node (value: int) =
    (* can declare self. self refers to itself (this in some languages), though could be named anything i.e. s
   that way we can call other methods within other methods using s#methodname format *)
    object (self)
        (* must declare as "mutable" if values can get updated,
           could have used "ref" type instead to accomplish the same along with
           dereference operators ! and := *)
        val mutable next = (None : node option)
        val mutable value = (value: int)
        (* setting up some custom setters/getters *)
        method getValue : int = value
        method getNext : node option = next
        method setNext (newNext: node option) : unit = next <- newNext
    end

class linked_list =
    object (self)
        val mutable head = (None : node option)
        method addLast (vle: int) =
            let newNode = new node vle in
                match head with
                    | None ->
                        head <- Some newNode
                    | Some hn ->
                        (* functions that can be recursive must be marked with rec *)
                        let rec goNext (current: node) =
                            let nextNode = current#getNext in
                                match nextNode with
                                    | None ->
                                        current#setNext (Some newNode)
                                    | Some m ->
                                        goNext m
                        in
                            goNext hn
        method addFirst (vle: int) : unit =
            let newNode = new node vle in
                match head with
                    | None ->
                        head <- Some newNode
                    | Some hn ->
                        newNode#setNext (Some hn);
                        head <- Some newNode
        method removeFirst : unit =
            match head with
                | None -> ()
                | Some h ->
                    head <- h#getNext
        method removeLast : unit =
            match head with
                | None -> ()
                | Some hn ->
                    let rec goNextNext (current: node) =
                        let nextNode = current#getNext in
                            match nextNode with
                                | None ->
                                    head <- None
                                | Some nn ->
                                    let nextNextNode = nn#getNext in
                                        match nextNextNode with
                                            | None ->
                                                current#setNext None
                                            | Some _ ->
                                                goNextNext nn
                    in
                        goNextNext hn
        method getFirst : int option =
            match head with
                | None -> None
                | Some h -> Some h#getValue
        method getLast : int option =
            match head with
                | None -> None
                | Some h ->
                    let rec goNext (current: node) =
                        let nextNode = current#getNext in
                            match nextNode with
                                | None ->
                                    Some current#getValue
                                | Some m ->
                                    goNext m
                    in
                        goNext h
        method display : unit =
            match head with
                | None -> () (* unit value constant, a not useful value to dispose *)
                | Some n ->
                    let rec goNext (current: node) =
                        let nextNode = current#getNext in
                            match nextNode with
                                | None ->
                                    Printf.printf "[%d]->\n" current#getValue;
                                | Some nn ->
                                    Printf.printf "[%d]->" current#getValue;
                                    goNext nn
                    in
                        goNext n
    end

let ll = new linked_list;;

ll#addLast 4;;
ll#addLast 3;;
ll#addLast 2;;
ll#addLast 1;;

ll#addFirst 5;;
ll#addFirst 6;;
ll#addFirst 7;;

ll#removeFirst;;
ll#removeLast;;

ll#display;;

let () =
    match ll#getFirst with
        | None -> Printf.printf "List is empty"
        | Some first ->
            Printf.printf "First Item: %d\n" first;

    match ll#getLast with
        | None -> Printf.printf "List is empty"
        | Some last ->
            Printf.printf "Last Item: %d\n" last;
